book/src/merklestatements.md

```diff
@@ -0,0 +1,197 @@
+
+## General derived statements for Merkle trees
+
+```
+MerkleSubtree(root: AnchoredKey::MerkleTree, node: AnchoredKey::MerkleTree)
+
+MerkleCorrectPath(root: AnchoredKey::MerkleTree, node: AnchoredKey::MerkleTree, key: AnchoredKey, depth: Value::Integer)
+
+Contains(root: AnchoredKey::MerkleTree, key: AnchoredKey, value: AnchoredKey)
+
+NotContains(root: AnchoredKey::MerkleTree, key: AnchoredKey)
+```
+
+### `MerkleSubtree`
+
+```
+MerkleSubtree(root, node)
+```
+means that there is a valid Merkle path of length `depth` from `root` to `node`.
+
+A `MerkleSubtree` statement is proved as follows:
+```
+MerkleSubtree(root, root)
+```
+is automatically true.
+
+Otherwise, `MerkleSubtree(root, node)` can be deduced from either
+```
+MerkleSubtree(root, parent)
+Branches(parent, node, other)
+```
+or
+```
+MerkleSubtree(root, parent)
+Branches(parent, other, node).
+```
+
+### `MerkleCorrectPath`
+
+```
+MerkleCorrectPath(root, node, key, depth)
+```
+means that there is a valid Merkle path of length `depth` from `root` to `node`, and if `key` appears as a key in the Merkle tree with root `root`, then `key` must be in the subtree under `node`.
+
+A `MerkleCorrectPath` statement is proved as follows:
+```
+MerkleCorrectPath(root, root, key, 0)
+```
+is automatically true.
+
+Otherwise, `MerkleCorrectPath(root, node, key, depth)` can be deduced from either:
+```
+MerkleCorrectPath(root, parent, key, depth-1)
+Branches(parent, node, other)
+GoesLeft(key, depth-1)
+```
+or
+```
+MerkleCorrectPath(root, parent, key, depth-1)
+Branches(parent, other, node)
+GoesRight(key, depth-1).
+```
+
+### `Contains`
+
+```
+Contains(root, key, value)
+```
+means that the key-value pair ```(key, value)``` is contained in the Merkle tree with Merkle root ```root```.
+
+A `Contains` statement can be deduced from the following two statements.
+```
+MerkleSubtree(root, node)
+Leaf(node, key, value)
+```
+
+### `NotContains`
+
+```
+NotContains(root, key)
+```
+means that the key ```key``` is not contained in the sparse Merkle tree with Merkle root ```root```.
+
+The statement `NotContains(root, key)` can be deduced from either
+```
+MerkleCorrectPath(root, node, key, depth)
+Leaf(node, otherkey, value)
+NotEqual(otherkey, key)
+```
+or
+```
+MerkleCorrectPath(root, node, key, depth)
+IsNullTree(node).
+```
+
+## Specialized statements for front-end compound types
+
+```
+ContainsHashedKey(root: AnchoredKey::DictOrSet, key: AnchoredKey)
+
+NotContainsHashedKey(root: AnchoredKey::DictOrSet, key: AnchoredKey)
+
+ContainsValue(root: AnchoredKey::Array, value: AnchoredKey)
+```
+
+When a dictionary or set is converted to a Merkle tree, its key is hashed -- see the [POD2 values page](./values.md#dictionary-array-set).
+
+```ContainsHashedKey(root, key)``` is deduced from
+```
+Contains(root, keyhash, value)
+keyhash = hash(key).
+```
+
+```NotContainsHashedKey(root, key)``` is deduced from 
+```
+NotContains(root, keyhash)
+keyhash = hash(key)
```

Comment:
I think we should have a custom statement for `Hash` :P
Then we could write all the high level merkle tree operations with custom statements.